lib/libzfs/libzfs_pool.c

```diff
@@ -4160,39 +4162,79 @@ zpool_get_errlog(zpool_handle_t *zhp, nvlist_t **nverrlistp)
 	verify(nvlist_alloc(nverrlistp, 0, KM_SLEEP) == 0);
 
 	/*
-	 * Fill in the nverrlistp with nvlist's of dataset and object numbers.
+	 * Fill in the nverrlistp with nvlist's of dataset number, object number
+	 * level position, and block id.
 	 */
+	nvlist_t *nv;
+	uint64_t *block_ids = zfs_alloc(zhp->zpool_hdl,
+	    count * sizeof (uint64_t));
+	int64_t *indrt_levels = zfs_alloc(zhp->zpool_hdl,
+	    count * sizeof (int64_t));
+	uint64_t error_count = 0;
 	for (i = 0; i < count; i++) {
-		nvlist_t *nv;
-
-		/* ignoring zb_blkid and zb_level for now */
-		if (i > 0 && zb[i-1].zb_objset == zb[i].zb_objset &&
-		    zb[i-1].zb_object == zb[i].zb_object)
+		if (i > 0 && zb[i - 1].zb_objset == zb[i].zb_objset &&
```

Comment:
I think it might be easier to read this loop if we the code appeared closer to the order that it's needed.

I also think it would help to group together all of the state related to the "current object".  This could be in a new struct, or by using the same prefix for these variables.  Using it's own struct would make it easier to reduce the duplicated code when we are adding this object's nvlist to nverrlistp.

Maybe something like:

```
for () {
  if (i == 0 || (different objset or object)) {
    if (i != 0) {
      // different logical object from the previous one.  Add nvlist for previous object.
      nvlist_add_nvlist(*nverrlistp...)
      nvlist_free(object_nv)
    }
    // new logical object.  Create nvlist for this one.
    object_nv = nvlist_alloc()
    ...
    object_error_count = 0;
  }
  // add this block to current object.
  object_blkids[object_error_count] = zb[i].zb_blkid;
  object_levels[object_error_count] = zb[i].zb_level;
  object_errors++
}
```